### DISCUSSION TITLE
Aura 4 Alpha: Briefing and Troubleshooting

### DISCUSSION THREAD

**Discussion by fosskers** [fosskers/aura#775] [General]
**Aura is an ALPM-based package manager and system management tool for Arch Linux and its derivatives.**

Hi everyone! :wave: Aura 4 (the Rust port) has been in development for some time, and is nearly ready for a release. Before announcing it on the most public channels, we're doing a limited Alpha and Beta to work out the initial post-port kinks. 

### Roadmap

- Alpha
  - All initial forms of major features.
  - English localisation.
- Beta
  - Fix Alpha bugs and oversights.
  - Port and expand localisation strings from Haskell Aura.
  - Shell completions.
  - Man page.
  - Rework _The Aura Book_.
  - Add more checks to `aura check`.
  - `crates.io` releases.
  - RFCs on Rust Discourse, This Week in Rust, etc.
- Release Candidates
  - Everything should work as expected by this point.
  - Final fixes and tweaks.
  - Solicit feedback from Arch-derivatives.

### Installation

`aura` can be installed directly from `crates.io`:
```
cargo install aura-pm --version 4.0.0-alpha3
```
Provided you have `~/.cargo/bin/` on your `PATH`, `aura` should now work:
```
❯ aura -V
aura 4.0.0-alpha3
```

Now, the first real thing you should check is `aura check`. **If there's something about its output that isn't clear to you, please ask about it here!**

### Usage and Expectations

The help menus have been vastly improved. Check out `aura -h` to see all the available subcommands (new or old), or paired with a subcommand like `aura -Ah`.

And a few other major usage notes:

- Aura now automatically asks for `sudo`, but only when needed. You _never_ need to prefix `aura` with `sudo`!
- Consequently, **pacman-like commands like `-Syu` no longer require `sudo`.** So: `aura -Syu` works just fine. 
- To upgrade AUR packages, use `-Au`. There are more subflags to add if you wish (check `aura -Ah`).
- Aura now uses TOML for configuration. The expected location is `~/.config/aura.toml`.
- Aura now keeps its various caches under `~/.cache/aura/`, not `/var/cache/aura/` as before.

Otherwise, as Aura is an ALPM-based package manager, it can do everything `pacman` can do. Go nuts.

### Questions / Discussion

- Can you think of anything to add to `aura check`?
- Should we compile with ALPM statically-linked?
- Does anything seem strange aesthetically?
- Have any of your `pacman` or Haskell Aura usage patterns broken?
- What's the output of `du -sh ~/.cargo/bin/aura` for you?
- Should Aura ban certain frowned-upon `pacman` commands like `-Rdd` and `-Su`?

**Comment by fosskers** [comment]
Aura can now be installed from `crates.io`!
```
cargo install aura-pm --version 4.0.0-alpha3
```

**Comment by name-tar-xz** [comment]
A few things
- `aura check` should probably check for non AUR or repo packages
- there should be an option in `~/.config/aura.toml` to change the sudo program used (like doas instead of sudo)

**Reply by fosskers** [reply]
1. That's a good idea and would be easy to implement.
2. Very reasonable, I've been planning to support `doas` for some time, but haven't yet gotten to it.

**Comment by etdr** [comment]
Quick question, would you say Aura 4 is a good reference for learning Rust in the same way that <4 was for Haskell?

**Reply by fosskers** [reply]
For learning how to write command line apps (especially ones with localisation!) I would say yes. I've worked hard to keep it simple and well-laid out.

**Comment by gagbo** [comment]
How do you deal with cycling dependencies ? I have an issue with mangohud here, which prevents any update.

```
aura :: AUR packages to upgrade:
 celt                        :: 0.11.3-5                   -> 0.11.3-6
 dxvk-bin                    :: 1.10.3-1                   -> 2.1-3
 epson-inkjet-printer-escpr  :: 1.7.21-1                   -> 1.7.24-1
 epson-inkjet-printer-escpr2 :: 1.1.49-1                   -> 1.1.60-1
 goverlay-bin                :: 0.9-1                      -> 0.9.1-2
 kasts                       :: 22.09.2-2                  -> 23.01.0-1
 libkipi                     :: 21.12.3-1                  -> 22.04.0-1
 lldb-mi-git                 :: r74.89945aa-1              -> r83.2388bd7-1
 mangohud                    :: 0.6.8-2                    -> 0.6.8-5
 mediawriter                 :: 5.0.1-1                    -> 5.0.4-1
 mu                          :: 1.8.9-1                    -> 1.8.14-1
 pomatez                     :: 1.1.0-1                    -> 1.2.2-2
 wezterm-git                 :: 20230121.182429.9350795f-1 -> 20230125.180646.f15bb186-1
 zenpower3-dkms              :: 0.2.0-1                    -> 0.2.0-3
aura :: Resolving dependencies...
aura :: Repository dependencies:
 rust
aura :: AUR packages:
 celt
 dxvk-bin
 epson-inkjet-printer-escpr
 epson-inkjet-printer-escpr2
 goverlay-bin
 kasts
 libkipi
 lldb-mi-git
 mangohud
 mediawriter
 mu
 pomatez
 wezterm-git
 zenpower3-dkms
aura :: Proceed? [Y/n] 
aura :: There was a cyclic dependency involving: mangohud
```

For the record:
[mangohud](https://aur.archlinux.org/packages/mangohud) seems to need [mangoapp](https://aur.archlinux.org/packages/mangoapp) as an optional dependency, which hard requires mangohud.

Is there an option to at least ignore that package so it updates anything that doesn't touch this cycle?

**Reply by fosskers** [reply]
I've [opened an issue](https://github.com/fosskers/aura/issues/802) for this.

**Comment by sonicbhoc** [comment]
I am also having an issue with cyclic dependencies.

I upgraded from Aura 3, which didn't have a problem with this, but after jumping to Aura 4.0.0-alpha8, I can no longer install pipewire-full-git, either by itself or with `aura -Ayu`.

Unfortunately with no depgraph or anything like that to look at, I can't see what aura sees and can't do much about it. And with no way to give aura instructions on how to resolve the dependency cycle, I can't upgrade the system with `aura -Ayu`.

**Reply by fosskers** [reply]
Thank you for reporting this. I've confirmed it: very strange that it resolves dependencies and offers to `Proceed`, but then fails as soon as you accept. 

I don't see anything out of the ordinary with `--log-level=debug`. 

Ah, here it is:
```rust
// Proceed if the user accepts.
    proceed!(fll, "proceed").ok_or(Error::Cancelled)?;

    // --- Determine the best build order --- //
    let order: Vec<Vec<&str>> = aura_core::aur::dependencies::build_order(&to_build)?;
    debug!("Build order: {:?}", order);
```
We see that build order is only determined after the user has accepted. Within `build_order` we see that the order is determined via a topological sort on the depgraph...
```rust
petgraph::algo::toposort(&graph, None)
        .map_err(|cycle| match graph.node_weight(cycle.node_id()) {
            None => Error::MalformedGraph,
            Some(b) => Error::CyclicDep(b.to_string()),
        })?
```

**Reply by fosskers** [reply]
> If the graph was acyclic, return a vector of nodes in topological order:
each node is ordered before its successors.
Otherwise, it will return a Cycle error. Self loops are also cycles.

Occasionally there are some funny self-dependencies in PKGBUILDs I suspect that may be the source of the issue here.

**Reply by fosskers** [reply]
Given how complex [this package definition is](https://aur.archlinux.org/cgit/aur.git/tree/.SRCINFO?h=pipewire-full-git), I suspect you will have to upgrade this package manually for now with `makepkg`.

**Reply by sonicbhoc** [reply]
What can be done to allow complex pkgbuilds like this to be installed by Aura? I'd like to be able to go back to using `aura -Ayu` to upgrade my system...

If there is code I can contribute I would not mind trying my hand at writing the required functionality.

**Reply by fosskers** [reply]
One _immediate_ option would be to add the `pipewire` packages to the `ignore` field of your `~/.config/aura.toml`. If you don't yet have one, you can generate one via:

```
aura conf --gen
```
the output of which you can copy to `~/.config/aura.toml`. This would at least allow you to `-Au` again freely.

Otherwise, the issue is how the "build order" is constructed. In the Rust, we're doing a naive Topological Sort of the dependency graph. In the ancient past, the original Haskell version used to do this too, but was eventually updated to do a number of manual graph operations and could handle complicated dep cycles like this. Unfortunately this logic hasn't yet been ported to the Rust.

If you'd like to try your hand at it, the file in question is `rust/aura-core/src/aur/dependencies.rs`. You can see the `toposort` occuring on Line 441. The corresponding Haskell starts in the `sortInstall` function, found in `haskell/aura/lib/Aura/Dependencies.hs`. Let me know if you have any luck with this!
